File: src/proj4/transformer.ts
import { NadGridStore } from '../readers/nadgrid';
import { parseProj } from './parseCode';
import { ALL_DEFINITIONS, DEFAULT_DEFINITIONS, EPSG_CODES, WGS84 } from './projections';
import { checkNotWGS, datumTransform } from './datum';

import type { MValue, VectorPoint, VectorPointM } from '../geometry';
import type {
  ProjectionParams,
  ProjectionTransform,
  ProjectionTransformDefinition,
} from './projections';

/**
 * # PROJ4 Transformer
 *
 * ## Description
 *
 * A Transformer class contains all projections necessary for converting coordinates from one
 * projection to another. This is a modular class that can be extended to add new projections
 * as needed to reduce code size and improve performance.
 * Both forward and inverse projections are default set to wgs84.
 *
 * Extends the {@link NadGridStore} class to support grid lookups
 *
 * ## Usage
 *
 * ### Full Example
 *
 * ```ts
 * import { Transformer, injectAllDefinitions, injectAllEPSGCodes } from 'gis-tools-ts';
 *
 * // Create a transform using a source and destination projection
 * const transform = new Transformer();
 * // inject all default definition projections. This is not memory efficient but ensures all
 * // projections are available
 * injectAllDefinitions(transform);
 * // inject all common EPSG codes. This is not memory efficient but ensures all EPSG codes are available
 * injectAllEPSGCodes(transform);
 * // If the transform requires a grid, this is how you add it.
 * transform.addGridFromReader(
 *   'BETA2007.gsb',
 *   new MMapReader(`${__dirname}/fixtures/BETA2007.gsb`),
 * );
 * // Set the source and destination projections
 * transform.setSource('EPSG_31466');
 * transform.setDestination('EPSG_25832');
 * // example forward projection
 * const forward = transform.forward({ x: 2559552, y: 5670982 });
 * // example inverse projection
 * const inverse = transform.inverse({ x: 349757.381712518, y: 5671004.06504954 });
 * ```
 *
 * ### Minimal Example only adding the Oblique Mercator
 *
 * ```ts
 * import { Transformer, HotineObliqueMercator, EPSG_8803 } from 'gis-tools-ts';
 *
 * const transform = new Transformer();
 * transform.insertDefinition(HotineObliqueMercator);
 * transform.insertEPSGCode('EPSG_8803', EPSG_8803);
 *
 * transform.setDestination('EPSG_8803');
 *
 * const forward = transform.forward({ x: 60.8, y: -132.2 });
 * ```
 */
export class Transformer extends NadGridStore {
  // EPSG code definitions
  epsgs = new Map<string, string>();
  // Definitions are descriptions of projections
  definitions = new Map<string, ProjectionTransformDefinition>();
  // source and destination projections
  wgs84: ProjectionTransform;
  source: ProjectionTransform;
  destination: ProjectionTransform;

  /**
   * Prepares default definitions, source transform, and destination transform
   * @param sourceCode - convenience: if provided, we run `this.setSource(sourceCode)` immediately
   * @param destCode - convenience: if provided, we run `this.setDestination(destCode)` immediately
   */
  constructor(sourceCode?: string | ProjectionParams, destCode?: string | ProjectionParams) {
    super();
    // by default supports the mercator and base (lon-lat) projection
    for (const def of DEFAULT_DEFINITIONS) this.insertDefinition(def);
    // defaults to a standard WGS84 lon-lat projection transform
    this.source = this.destination = this.wgs84 = this.#buildTransformer(WGS84);
    if (sourceCode !== undefined) this.setSource(sourceCode);
    if (destCode !== undefined) this.setDestination(destCode);
  }

  /**
   * Set the source projection
   * @param sourceCode - can be a name or a coded definition
   */
  setSource(sourceCode: string | ProjectionParams): void {
    this.source = this.#buildTransformer(sourceCode);
  }

  /**
   * Set the destination projection
   * @param destCode - can be a name or a coded definition
   */
  setDestination(destCode: string | ProjectionParams): void {
    this.destination = this.#buildTransformer(destCode);
  }

  /**
   * Build a ProjectionTransform
   * @param code - can be a WKT object or proj4 encoded string
   * @returns - A ready to use ProjectionTransform
   */
  #buildTransformer(code: string | ProjectionParams): ProjectionTransform {
    code = this.#parseEPSGCode(code);
    const params = parseProj(code, this);
    // search
    let def: ProjectionTransformDefinition | undefined;
    for (const name of [params.projName, params.name]) {
      def = this.definitions.get(name?.toLowerCase() ?? '');
      if (def !== undefined) break;
    }
    if (def === undefined)
      throw Error(`${params.projName ?? params.name} invalid, unsupported, or not loaded`);
    return new def(params);
  }

  /**
   * Attempt to parse the EPSG code if epsg codes are loaded and return the definition
   * @param code - a string or params with the EPSG code set in projName
   * @returns - if no EPSG code is found, returns the original code. Otherwise returns the EPSG definition
   */
  #parseEPSGCode(code: string | ProjectionParams): string | ProjectionParams {
    const codeName = (typeof code === 'string' ? code : code.projName)?.replaceAll(':', '_');
    const epsg = this.epsgs.get(codeName ?? '');
    if (epsg !== undefined) return epsg;
    return code;
  }

  /**
   * Insert a projection definition
   * ```ts
   * import { Transformer, HotineObliqueMercator } from 'gis-tools-ts';
   * const transformer = new Transformer();
   * transformer.insertDefinition(HotineObliqueMercator);
   * ```
   * @param def - a class that may be instatiated with future setSource and setDestination
   * @param names - optionally add projection reference names to add lookups to the definition
   */
  insertDefinition(def: ProjectionTransformDefinition, names: string[] = []): void {
    for (const name of def.names) this.definitions.set(name.toLowerCase(), def);
    for (const name of names) this.definitions.set(name.toLowerCase(), def);
  }

  /**
   * Insert an EPSG code definition
   * ```ts
   * import { Transformer, EPSG_4326 } from 'gis-tools-ts';
   * const transformer = new Transformer();
   * transformer.insertEPSGCode('EPSG_4326', EPSG_4326);
   * ```
   * @param code - EPSG code to insert e.g. "EPSG_4326" (uses underscore instead of colon)
   * @param value - the EPSG definition which is either a WKT string object or proj4 encoded string
   */
  insertEPSGCode(code: string, value: string): void {
    this.epsgs.set(code, value);
  }

  /**
   * Forward projection from src projection to dest projection
   * ```ts
   * const transformer = new Transformer();
   * transformer.setSource('EPSG_4326');
   * const point = transformer.forward({ x: 0, y: 0 });
   * ```
   * @param p - vector point currently in the "source" projection
   * @param enforceAxis - enforce axis ensures axis consistency relative to the final projection
   * @returns - vector point in the "destination" projection
   */
  forward<D extends MValue>(p: VectorPoint<D>, enforceAxis?: false): VectorPoint<D>;
  /**
   * Forward projection from src projection to dest projection
   * ```ts
   * const transformer = new Transformer();
   * transformer.setSource('EPSG_4326');
   * const point = transformer.forward({ x: 0, y: 0 });
   * ```
   * @param p - vector point currently in the "source" projection
   * @param enforceAxis - enforce axis ensures axis consistency relative to the final projection
   * @returns - vector point in the "destination" projection
   */
  forward<D extends MValue>(p: VectorPointM<D>, enforceAxis?: false): VectorPointM<D>;

  /**
   * Forward projection from src projection to dest projection
   * ```ts
   * const transformer = new Transformer();
   * transformer.setSource('EPSG_4326');
   * const point = transformer.forward({ x: 0, y: 0 });
   * ```
   * @param p - vector point currently in the "source" projection
   * @param enforceAxis - enforce axis ensures axis consistency relative to the final projection
   * @returns - vector point in the "destination" projection
   */
  forward<D extends MValue>(
    p: VectorPoint<D> | VectorPointM<D>,
    enforceAxis = false,
  ): VectorPoint<D> | VectorPointM<D> {
    return this.#transformPoint(p, this.source, this.destination, enforceAxis);
  }

  /**
   * Inverse projection from dest projection to src projection
   * ```ts
   * const transformer = new Transformer();
   * transformer.setSource('EPSG_4326');
   * const point = transformer.inverse({ x: 0, y: 0 });
   * ```
   * @param p - vector point currently in the "destination" projection
   * @param enforceAxis - enforce axis ensures axis consistency relative to the final projection
   * @returns - vector point in the "source" projection
   */
  inverse<D extends MValue>(p: VectorPoint<D>, enforceAxis = false): VectorPoint<D> {
    return this.#transformPoint(p, this.destination, this.source, enforceAxis);
  }

  /**
   * Transforms a point from one projection to another
   * @param sourcePoint - point to start transforming
   * @param src - source projection
   * @param dest - destination projection
   * @param enforceAxis - enforce axis ensures axis consistency relative to the final projection
   * @returns - transformed point
   */
  #transformPoint<M extends MValue>(
    sourcePoint: VectorPoint<M>,
    src: ProjectionTransform,
    dest: ProjectionTransform,
    enforceAxis: boolean,
  ): VectorPoint<M> {
    // Workaround for datum shifts towgs84, if either source or destination projection is not wgs84
    let res: VectorPoint<M> = { ...sourcePoint };
    if (src === dest) return res;
    const hasZ = sourcePoint.z !== undefined;
    if (checkNotWGS(src, dest)) {
      const wgs84 = this.wgs84;
      res = this.#transformPoint(res, src, wgs84, enforceAxis);
      src = wgs84;
    }
    // STEP 1: SOURCE -> WGS84
    // if needed, adjust axis
    if (enforceAxis && src.axis !== 'enu') adjustAxis(res, src, true);
    // adjust for meters if necessary
    if (src.name !== 'longlat' && src.projName !== 'longlat' && src.toMeter !== undefined) {
      res.x *= src.toMeter;
      res.y *= src.toMeter;
    }

    // transform forward
    src.inverse(res);
    // Adjust for the prime meridian if necessary
    res.x += src.fromGreenwich;

    // STEP 2: MID-POINT. Convert datums if needed, and if possible.
    datumTransform(res, src, dest);

    // STEP 3: WGS84 -> DEST
    // Adjust for the prime meridian if necessary
    res.x -= dest.fromGreenwich;
    // transform forward
    dest.forward(res);
    // adjust for meters if necessary
    if (dest.name !== 'longlat' && dest.projName !== 'longlat' && dest.toMeter !== undefined) {
      res.x /= dest.toMeter;
      res.y /= dest.toMeter;
    }
    // if needed, adjust axis
    if (enforceAxis && dest.axis !== 'enu') adjustAxis(res, dest, true);

    if (!hasZ) delete res.z;

    return res;
  }
}

/**
 * Inject all default definitions into the transformer
 * @param transformer - projection transformer
 */
export function injectAllDefinitions(transformer: Transformer): void {
  for (const proj of ALL_DEFINITIONS) transformer.insertDefinition(proj);
}

/**
 * Inject all EPSG codes into the transformer
 * @param transformer - the transformer to inject EPSG codes to
 */
export function injectAllEPSGCodes(transformer: Transformer): void {
  for (const [key, value] of Object.entries(EPSG_CODES)) transformer.insertEPSGCode(key, value);
}

/**
 * Adjusts an axis
 * @param point - the vector point to adjust
 * @param prj - the projection
 * @param denorm - denormalizes z if true
 */
function adjustAxis<M extends MValue>(
  point: VectorPoint<M>,
  prj: ProjectionTransform,
  denorm: boolean,
): void {
  const xin = point.x;
  const yin = point.y;
  const zin = point.z ?? 0;
  let v, i;
  let t: 'x' | 'y' | 'z';
  for (i = 0; i < 3; i++) {
    if (denorm && i === 2 && point.z === undefined) continue;
    if (i === 0) {
      v = xin;
      if ('ew'.indexOf(prj.axis[i]) !== -1) {
        t = 'x';
      } else {
        t = 'y';
      }
    } else if (i === 1) {
      v = yin;
      if ('ns'.indexOf(prj.axis[i]) !== -1) {
        t = 'y';
      } else {
        t = 'x';
      }
    } else {
      v = zin;
      t = 'z';
    }
    switch (prj.axis[i]) {
      case 'e':
        point[t] = v;
        break;
      case 'w':
        point[t] = -v;
        break;
      case 'n':
        point[t] = v;
        break;
      case 's':
        point[t] = -v;
        break;
      case 'u':
        if (point[t] !== undefined) {
          point.z = v;
        }
        break;
      case 'd':
        if (point[t] !== undefined) {
          point.z = -v;
        }
        break;
      default:
        throw Error(`unknown axis (${prj.axis[i]}) - check definition of ${prj.name}`);
    }
  }
}

File: rust/proj/transform/transformer.rs
use super::TransformCoordinates;
use crate::proj::{
    Direction, IoUnits, ProjJSON, ProjectionTransform, adjlon, check_not_wgs84, datum_transform,
    geocentric_latitude,
};
use alloc::{collections::BTreeMap, fmt::Debug, string::String};
use core::f64::consts::FRAC_PI_2;

/// # PROJ Transformer
///
/// ## Description
///
/// A Transformer class contains all projections necessary for converting coordinates from one
/// projection to another. This is a modular class that can be extended to add new projections
/// as needed to reduce code size and improve performance.
/// Both forward and inverse projections are default set to wgs84.
///
/// See the NadGridStore
///
/// ## Usage
///
/// ### Full Example
///
/// ```ts
/// // TODO
/// ```
#[derive(Debug, Default, Clone, PartialEq)]
pub struct Transformer {
    epsgs: BTreeMap<String, String>,
    src: ProjectionTransform,
    dest: ProjectionTransform,
}
impl Transformer {
    /// Create a new Transformer
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self {
            epsgs: BTreeMap::new(),
            src: ProjectionTransform::wgs84(),
            dest: ProjectionTransform::wgs84(),
        }
    }
    /// Move forward from source projection to destination projection
    pub fn forward<P: TransformCoordinates + Debug>(&self, p: &P) -> P {
        let mut res = p.clone();
        transform_point(&self.src, &self.dest, &mut res);
        res
    }
    /// Move forward from source projection to destination projection in place
    pub fn forward_mut<P: TransformCoordinates + Debug>(&self, p: &mut P) {
        transform_point(&self.src, &self.dest, p);
    }
    /// Move backward from destination projection to source projection
    pub fn inverse<P: TransformCoordinates + Debug>(&self, p: &P) -> P {
        let mut res = p.clone();
        transform_point(&self.dest, &self.src, &mut res);
        res
    }
    /// Move backward from destination projection to source projection in place
    pub fn inverse_mut<P: TransformCoordinates + Debug>(&self, p: &mut P) {
        transform_point(&self.dest, &self.src, p);
    }

    /// Insert an EPSG code definition
    /// ```ts
    /// // TODO
    /// ```
    /// @param code - EPSG code to insert e.g. "4326": "WKT_STRING"
    /// @param value - the EPSG definition which is either a WKT string object or proj4 encoded string
    pub fn insert_epsg_code(&mut self, code: String, value: String) {
        self.epsgs.insert(code, value);
    }

    /// Set the source projection
    /// @param code - can be a name or a json/wkt coded definition
    pub fn set_source(&mut self, code: String) {
        self.src = self.build_transformer(code);
    }

    /// Set the source projection
    /// @param def - transform definition
    pub fn set_source_def(&mut self, def: ProjectionTransform) {
        self.src = def;
    }

    /// Set the destination projection
    /// @param code - can be a name or a coded definition
    pub fn set_destination(&mut self, code: String) {
        self.dest = self.build_transformer(code);
    }

    /// Set the source projection
    /// @param def - transform definition
    pub fn set_destination_def(&mut self, def: ProjectionTransform) {
        self.dest = def;
    }

    /// Build a ProjectionTransform
    /// @param code - can be a WKT object or proj4 encoded string
    /// @returns - A ready to use ProjectionTransform
    fn build_transformer(&mut self, mut code: String) -> ProjectionTransform {
        if let Some(epsg) = self.epsgs.get(&code) {
            code = epsg.clone();
        }
        // TRY JSON
        if let Ok(json) = serde_json::from_str::<ProjJSON>(&code) {
            return json.to_projection_transform();
        }
        // TRY WKT
        ProjJSON::parse_wkt(&code).to_projection_transform()
    }

    /// Get access to an epsg code
    pub fn get_epsg_code(&self, code: String) -> Option<String> {
        self.epsgs.get(&code).cloned()
    }
}

/// Transforms a point from one projection to another
/// @param src - source projection
/// @param dest - destination projection
/// @param point - point to mutate
pub fn transform_point<P: TransformCoordinates + Debug>(
    src: &ProjectionTransform,
    dest: &ProjectionTransform,
    point: &mut P,
) {
    // Check if there are any steps
    if src == dest || (src.is_wgs84 && dest.is_wgs84) {
        return;
    }
    let has_z = point.has_z();

    // TODO: We ideally drop this when cart and cart_wgs84 are merged
    // STEP 0: If the datums are incompatible, be sure to use the intermediate wgs84 transform
    if check_not_wgs84(src, dest) {
        transform_point(src, &ProjectionTransform::wgs84(), point);
    }

    // STEP 1 INVERSE
    transform_inv(src, point);
    // STEP 2: MID-POINT. Convert datums if needed, and if possible.
    datum_transform(point, &src.proj.borrow(), &dest.proj.borrow());
    // STEP 3: FORWARD
    transform_fwd(dest, point);

    if !has_z {
        point.set_z(0.0);
    }
}

fn transform_inv<P: TransformCoordinates + Debug>(
    proj_trans: &ProjectionTransform,
    coords: &mut P,
) {
    transform_inv_prepare(proj_trans, coords);
    proj_trans.method.inverse(coords);
    transform_inv_finalize(proj_trans, coords);
}

fn transform_inv_prepare<P: TransformCoordinates + Debug>(
    proj_trans: &ProjectionTransform,
    coords: &mut P,
) {
    let proj = &proj_trans.proj.borrow();
    // The helmert datum shift will choke unless it gets a sensible 4D coordinate
    if f64::INFINITY == coords.z() && proj_trans.helmert.is_some() {
        coords.set_z(0.0);
    }
    if f64::INFINITY == coords.t() && proj_trans.helmert.is_some() {
        coords.set_t(0.0);
    }

    if let Some(axisswap) = &proj_trans.axisswap {
        axisswap.method.inverse(coords);
    }

    // Handle remaining possible input types
    match proj.right {
        // de-scale and de-offset
        IoUnits::CARTESIAN => {
            let to_meter = proj.to_meter;
            coords.set_x(coords.x() * to_meter);
            coords.set_y(coords.y() * to_meter);
            coords.set_z(coords.z() * to_meter);
            if proj.is_geocent
                && let Some(cart) = &proj_trans.cart
            {
                transform_inv(cart, coords);
            }
        }

        IoUnits::PROJECTED | IoUnits::CLASSIC => {
            let to_meter = proj.to_meter;
            let vto_meter = proj.vto_meter;
            coords.set_x(coords.x() * to_meter - proj.x0);
            // 440720.0 - 400000
            coords.set_y(coords.y() * to_meter - proj.y0);
            coords.set_z(coords.z() * vto_meter - proj.z0);
            if proj.right == IoUnits::PROJECTED {
                return;
            }

            // Classic proj.4 functions expect plane coordinates in units of the semimajor axis
            // Multiplying by ra, rather than dividing by a because the CalCOFI projection
            // stomps on a and hence (apparently) depends on this to roundtrip correctly
            // (CalCOFI avoids further scaling by stomping - but a better solution is possible)
            if proj.ra != 0. {
                coords.set_x(coords.x() * proj.ra);
                coords.set_y(coords.y() * proj.ra);
            }
        }

        IoUnits::RADIANS => {
            coords.set_z(coords.z() * proj.vto_meter - proj.z0);
        }
        _ => {}
    }
}

fn transform_inv_finalize<P: TransformCoordinates + Debug>(
    proj_trans: &ProjectionTransform,
    coords: &mut P,
) {
    let proj = &proj_trans.proj.borrow();
    if proj.left == IoUnits::RADIANS {
        // Distance from central meridian, taking system zero meridian into account
        coords.set_lam(coords.lam() + proj.from_greenwich + proj.lam0);

        // adjust longitude to central meridian
        if !proj.over {
            coords.set_lam(adjlon(coords.lam()));
        }

        if let Some(vgridshift) = &proj_trans.vgridshift {
            // Go geometric from orthometric
            transform_inv(vgridshift, coords);
        }
        if let Some(hgridshift) = &proj_trans.hgridshift {
            // Go geometric from orthometric
            transform_fwd(hgridshift, coords);
        } else if (proj_trans.cart_wgs84.is_some() && proj_trans.cart.is_some())
            || proj_trans.helmert.is_some()
        {
            // Go cartesian in local frame
            if let Some(cart) = &proj_trans.cart {
                transform_fwd(cart, coords);
            }
            // Step into WGS84
            if let Some(helmert) = &proj_trans.helmert {
                transform_fwd(helmert, coords);
            }
            // Go back to angular using WGS84 ellps
            if let Some(cart_wgs84) = &proj_trans.cart_wgs84 {
                transform_inv(cart_wgs84, coords);
            }
        }

        // If input latitude was geocentrical, convert back to geocentrical
        if proj.geoc {
            geocentric_latitude(proj, Direction::FWD, coords);
        }
    }
}

fn transform_fwd<P: TransformCoordinates + Debug>(
    proj_trans: &ProjectionTransform,
    coords: &mut P,
) {
    transform_fwd_prepare(proj_trans, coords);
    proj_trans.method.forward(coords);
    transform_fwd_finalize(proj_trans, coords);
}

fn transform_fwd_prepare<P: TransformCoordinates + Debug>(
    proj_trans: &ProjectionTransform,
    coords: &mut P,
) {
    let proj = &proj_trans.proj.borrow();
    // The helmert datum shift will choke unless it gets a sensible 4D coordinate
    if coords.z() == f64::INFINITY && proj_trans.helmert.is_some() {
        coords.set_z(0.0);
    }
    if coords.t() == f64::INFINITY && proj_trans.helmert.is_some() {
        coords.set_t(0.0);
    }

    // Check validity of angular input coordinates
    if proj.left == IoUnits::RADIANS {
        // check for latitude or longitude over-range
        let t = (if coords.phi() < 0. { -coords.phi() } else { coords.phi() }) - FRAC_PI_2;
        if t > 1e-12 || coords.lam() > 10. || coords.lam() < -10. {
            panic!("Invalid latitude");
        }
        // Clamp latitude to -90..90 degree range
        if coords.phi() > FRAC_PI_2 {
            coords.set_phi(FRAC_PI_2);
        }
        if coords.phi() < -FRAC_PI_2 {
            coords.set_phi(-FRAC_PI_2);
        }
        // If input latitude is geocentrical, convert to geographical */
        if proj.geoc {
            geocentric_latitude(proj, Direction::INV, coords);
        }
        // Ensure longitude is in the -pi:pi range
        if !proj.over {
            coords.set_lam(adjlon(coords.lam()));
        }

        if let Some(hgridshift) = &proj_trans.hgridshift {
            transform_inv(hgridshift, coords);
        } else if (proj_trans.cart_wgs84.is_some() && proj_trans.cart.is_some())
            || proj_trans.helmert.is_some()
        {
            // Go cartesian in local frame
            if let Some(cart_wgs84) = &proj_trans.cart_wgs84 {
                transform_fwd(cart_wgs84, coords);
            }
            // Step into WGS84
            if let Some(helmert) = &proj_trans.helmert {
                transform_inv(helmert, coords);
            }
            // Go back to angular using WGS84 ellps
            if let Some(cart) = &proj_trans.cart {
                transform_inv(cart, coords);
            }
        }
        // Go orthometric from geometric
        if let Some(vgridshift) = &proj_trans.vgridshift {
            transform_fwd(vgridshift, coords);
        }
        // Distance from central meridian, taking system zero meridian into account
        coords.set_lam((coords.lam() - proj.from_greenwich) - proj.lam0);

        // Ensure longitude is in the -pi:pi range
        if !proj.over {
            coords.set_lam(adjlon(coords.lam()));
        }

        return;
    }

    // We do not support gridshifts on cartesian input
    if proj.left == IoUnits::CARTESIAN
        && let Some(helmert) = &proj_trans.helmert
    {
        transform_inv(helmert, coords);
    }
}

fn transform_fwd_finalize<P: TransformCoordinates + Debug>(
    proj_trans: &ProjectionTransform,
    coords: &mut P,
) {
    let proj = &proj_trans.proj.borrow();
    match proj.right {
        // Handle false eastings/northings and non-metric linear units
        IoUnits::CARTESIAN => {
            if proj.is_geocent
                && let Some(cart) = &proj_trans.cart
            {
                transform_fwd(cart, coords);
            }
            coords.set_x(coords.x() * proj.fr_meter);
            coords.set_y(coords.y() * proj.fr_meter);
            coords.set_z(coords.z() * proj.vfr_meter);
        }

        // Classic proj.4 functions return plane coordinates in units of the semimajor axis
        IoUnits::CLASSIC => {
            if proj.a != 0. {
                coords.set_x(coords.x() * proj.a);
                coords.set_y(coords.y() * proj.a);
            }
            coords.set_x(proj.fr_meter * (coords.x() + proj.x0));
            coords.set_y(proj.fr_meter * (coords.y() + proj.y0));
            coords.set_z(proj.vfr_meter * (coords.z() + proj.z0));
        }

        // to continue processing in common with IoUnits::PROJECTED
        IoUnits::PROJECTED => {
            coords.set_x(proj.fr_meter * (coords.x() + proj.x0));
            coords.set_y(proj.fr_meter * (coords.y() + proj.y0));
            coords.set_z(proj.vfr_meter * (coords.z() + proj.z0));
        }

        IoUnits::RADIANS => {
            coords.set_z(proj.vfr_meter * (coords.z() + proj.z0));
            // not interested in adding this support. keeping for posterity
            // if (proj.is_long_wrap_set) {
            //     if (coordsz.lam != HUGE_VAL) {
            //         coordsz.lam =
            //             proj.long_wrap_center + adjlon(coordsz.lam - proj.long_wrap_center);
            //     }
            // }
        }
        _ => {}
    }

    if let Some(axisswap) = &proj_trans.axisswap {
        axisswap.method.forward(coords);
    }
}
